src/LEPTON/fix_epot_lepton.cpp

```diff
@@ -0,0 +1,322 @@
+
+void FixEpotLepton::post_force(int vflag)
+{
+  double **f = atom->f;
+  double **x = atom->x;
+  double *q = nullptr;
+  double **mu = nullptr, **t = nullptr;
+  int *mask = atom->mask;
+  imageint *image = atom->image;
+  int nlocal = atom->nlocal;
+
+  const double qe2f = force->qe2f;
+  const double qqr2e = force->qqr2e;
+
+  auto parsed = Lepton::Parser::parse(LeptonUtils::substitute(expr, lmp)).optimize();
+  Lepton::CompiledExpression phi, dphi_x, dphi_y, dphi_z, dphi_xx, dphi_xy, dphi_xz, dphi_yy, dphi_yz, dphi_zz; 
+
+  dphi_x = parsed.differentiate("x").createCompiledExpression();
+  dphi_y = parsed.differentiate("y").createCompiledExpression();
+  dphi_z = parsed.differentiate("z").createCompiledExpression();
+
+  std::vector<Lepton::CompiledExpression*> cexprs = {&dphi_x, &dphi_y, &dphi_z};
+
+  if (atom->q_flag) {
+    q = atom->q;
+    phi = parsed.createCompiledExpression();
+    cexprs.push_back(&phi);
+  }
+  if (atom->mu_flag) {
+    mu = atom->mu; t = atom->torque;
+    dphi_xx = parsed.differentiate("x").differentiate("x").createCompiledExpression();
+    dphi_xy = parsed.differentiate("x").differentiate("y").createCompiledExpression();
+    dphi_xz = parsed.differentiate("x").differentiate("z").createCompiledExpression();
+    dphi_yy = parsed.differentiate("y").differentiate("y").createCompiledExpression();
+    dphi_yz = parsed.differentiate("y").differentiate("z").createCompiledExpression();
+    dphi_zz = parsed.differentiate("z").differentiate("z").createCompiledExpression();
+    cexprs.insert(cexprs.end(), {&dphi_xx, &dphi_xy, &dphi_xz, &dphi_yy, &dphi_yz, &dphi_zz}); 
+  }
+
+  // Substitute x, y, z if they exist
+  const std::array<std::string, 3> variableNames = {"x", "y", "z"}; 
+  std::vector<std::array<bool, 3>> has_ref;     
+  for (auto &cexpr : cexprs) {
+    has_ref.push_back({true, true, true});
+    for (size_t i = 0; i < variableNames.size(); i++) {
+      try { 
+        (*cexpr).getVariableReference(variableNames[i]);
+      }
+      catch (Lepton::Exception &) {
+        has_ref.back()[i] = false;
+      }
+    }
+  }
+  // virial setup
+  v_init(vflag);
+
+  // update region if necessary
+  if (region) region->prematch();
+
+  // fsum[0] = "potential energy" for added force
+  // fsum[123] = extra force added to atoms
+  fsum[0] = fsum[1] = fsum[2] = fsum[3] = 0.0;
+  force_flag = 0;
+
+  double fx, fy, fz;
+  double v[6], unwrap[3];
+
+  for (int i = 0; i < nlocal; i++) {
+    if (mask[i] & groupbit) {
+      if (region && !region->match(x[i][0], x[i][1], x[i][2])) continue;
+      domain->unmap(x[i], image[i], unwrap);
+      
+      // Substitute x, y, z if they exist           
+      for (size_t j = 0; j< cexprs.size(); j++) {
+        if (has_ref[j][0]) (*cexprs[j]).getVariableReference("x") = unwrap[0];
+        if (has_ref[j][1]) (*cexprs[j]).getVariableReference("y") = unwrap[1];
+        if (has_ref[j][2]) (*cexprs[j]).getVariableReference("z") = unwrap[2];
+      }
```

Comment:
If I understand correctly this imposes forces calculated from _unwrapped_ coordinates, with lots of possible unintended consequences -- for example a quadratic field will impose a different force on charges depending on their image flag.

I would suggest enabling users to write their Lepton expressions with a wrapped `x` co-ordinate vs an unwrapped `ux` co-ordinate (= x + ix*lx -- which is itself problematic in NPT sims), or whatever equivalent is available in Lepton parsing.

It's messy either way -- then again, inhomogeneous electric fields in periodic simulations ARE inherently messy 😈